Prototypal inheritance (#207)
javascript-tutorial/it.javascript.info: 1-js/08-prototypes/01-prototype-inheritance/article.md

# Prototypal inheritance

> Nella programmazione, spesso vogliamo prendere qualcosa ed estenderlo.

Reviewer: ```suggestion
Nella programmazione, spesso vogliamo prendere qualcosa ed estenderla.
```

Ad esempio, potremmo avere un oggetto `user` con le sue proprietà e i suoi metodi, e voler definire gli oggetti `admin` e `guest` come sue varianti. Vorremmo però poter riutilizzare ciò che abbiamo nell'oggetto `user`, evitando di copiare e reimplementare nuovamente i suoi metodi, quindi vorremmo semplicemente definire un nuovo oggetto a partire da esso.

La *prototypal inheritance* (ereditarietà dei prototype) è una caratteristica del linguaggio che aiuta in questo senso.

## [[Prototype]]

> In JavaScript, possiedono una speciale proprietà nascosta `[[Prototype]]` (come definito nella specifica), questo può valere `null` oppure può contenere il riferimento ad un altro oggetto. Quell'oggetto viene definito "prototype" (prototipo):

Reviewer: ```suggestion
In JavaScript, gli oggetti possiedono una speciale proprietà nascosta `[[Prototype]]` (come definito nella specifica); questo può valere `null` oppure può contenere il riferimento ad un altro oggetto. Quell'oggetto viene definito "prototype" (prototipo):
```

![prototype](object-prototype-empty.svg)

> Quando leggiamo una proprietà da `object`, e questa non esiste, JavaScript prova automaticamente a recuperarla dal suo prototype. In programmazione, questo comportamento viene definito "prototypal inheritance". Presto vederemo diversi esempi di questo tipo di ereditarietà, e vedremo anche delle interessanti caratteristiche di linguaggio basate su di essa. 

Reviewer: ```suggestion
Quando leggiamo una proprietà da `object`, e questa non esiste, JavaScript prova automaticamente a recuperarla dal suo prototype. In programmazione, questo comportamento viene definito "prototypal inheritance". Presto vederemo diversi esempi di questo tipo di ereditarietà, e vedremo anche delle interessanti caratteristiche del linguaggio basate su di essa. 
```

La proprietà `[[Prototype]]` è interna e nascosta, ma esistono diversi modi per poterla impostare.

> Uno di questi è quello di utilizzare la nomenclatura speciale `__proto__`, in questo modo:

Reviewer: ```suggestion
Uno di questi è quello di utilizzare la nomenclatura speciale `__proto__`:
```
toglierei l'ultima parte, che ne dici?

};

*!*

> rabbit.__proto__ = animal; // imposta il prototy di rabbit,.[[Prototype]] = animal

Reviewer: ```suggestion
rabbit.__proto__ = animal; // imposta il prototype di rabbit,.[[Prototype]] = animal
```

rabbit.__proto__ = animal; // (*)
*/!*

// ora in rabbit possiamo trovare entrambe le proprietà
*!*
alert( rabbit.eats ); // true (**)
*/!*
alert( rabbit.jumps ); // true
```

Nell'esempio la linea `(*)` imposta `animal` come prototype di `rabbit`.

Successivamente, quando `alert` proverà a leggere la proprietà `rabbit.eats` `(**)`, non la troverà in rabbit, quindi JavaScript seguirà il riferimento in `[[Prototype]]` e la troverà in  `animal` (ricerca dal basso verso l'alto):

![](proto-animal-rabbit.svg)

> In questo caso possiamo dire che "`animal` è il prototype di `rabbit`" o, in alternativa, che "`rabbit` prototypically inherits (eredità dal prototipo) da `animal`"

Reviewer: ```suggestion
In questo caso possiamo dire che "`animal` è il prototype di `rabbit`" o, in alternativa, che "`rabbit` *prototypically inherits* (eredità dal prototipo) da `animal`"
```
In questo caso il corsivo ci sta :)

*/!*
};

// walk viene presa dalla catena di prototype
longEar.walk(); // Animal walk
alert(longEar.jumps); // true (da rabbit)
```

![](proto-animal-rabbit-chain.svg)

Ora, se provassimo a leggere qualcosa da `longEar`, e non esistesse, JavaScript andrebbe a guardare prima in `rabbit`, e poi in `animal`.

Ci sono solamente due limitazioni:

> 1. Non possono esserci riferimenti circolari. JavaScript lancerebbe un errore se provassimo ad assegnare a `__proto__` un riferimento circolare.

Reviewer 2: ```suggestion
1. Non possono esserci riferimenti circolari. JavaScript restituirebbe un errore se provassimo ad assegnare a `__proto__` un riferimento circolare.
```

> 2. Il valore di `__proto__` può essere o un oggetto o `null`. Gli altri valore vengono ignorati.

Reviewer: ```suggestion
2. Il valore di `__proto__` può essere o un oggetto o `null`. Gli altri valori vengono ignorati.
```

> Inoltre, anche se dovrebbe essere già ovvio: può esserci solamente un `[[Prototype]]`. Un oggetto non può eridatare da più oggetti.

Reviewer: ```suggestion
Inoltre, anche se dovrebbe essere già ovvio: può esserci solamente un `[[Prototype]]`. Un oggetto non può ereditare da più oggetti.
```

```smart header="`__proto__` è un getter/setter storico per `[[Prototype]]`"
E' un errore comune tra i principianti quello di non conoscere la differenza tra questi due.

> Da notare che `__proto__` non è *la stessa cosa* della proprietà `[[Prototype]]`. E' solamente un getter/setter per `[[Prototype]]`. Più avanti vedremo alcune situazioni in cui questa differenza avrà importanza, quando avremo una buona conoscenza del linguaggio JavaScript.

Reviewer 2: ```suggestion
Da notare che `__proto__` non è *la stessa cosa* della proprietà `[[Prototype]]`. E' solamente un getter/setter per `[[Prototype]]`. Più avanti vedremo alcune situazioni in cui questa differenza avrà importanza, ma per ora tenetelo solo a mente.
```

La proprietà `__proto__` è leggermente datata. Esiste solamente per ragioni storiche, la versione attuale di JavaScript suggerisce di utilizzare le funzioni `Object.getPrototypeOf/Object.setPrototypeOf` per impostare il prototype. Vedremo meglio queste funzioni più avanti.

Secondo la specifica, `__proto__` deve essere supportato solamente dai browser. In realtà, tutti gli ambienti, inclusi quelli server-side, supportano `__proto__`, quindi il suo utilizzo è piuttosto sicuro.

> Poichè la notazione `__proto__` risulta essere più intuitiva, la utilizzeremo nei nostri esempi.

Reviewer: ```suggestion
Poichpoiché la notazione `__proto__` risulta essere più intuitiva, la utilizzeremo nei nostri esempi.
```

```

## La scrittura non utilizza prototype

Il prototype viene utilizzato solamente per la lettura delle proprietà.

Le operazioni di scrittura/rimozione utilizzano direttamente l'oggetto.

> Nell'esempio che vediamo sotto, assegniamo un suo metodo `walk` a `rabbit`:

Reviewer: ```suggestion
Nell'esempio che vediamo sotto, assegniamo un metodo `walk` a `rabbit`, che sarà solo suo:
```
Mi sembra piu' chiaro cosi', che ne pensi?

alert(admin.fullName); // John Smith (*)

// il setter viene invocato!
admin.fullName = "Alice Cooper"; // (**)

alert(admin.fullName); // Alice Cooper, lo stato di admin è stato modificato
alert(user.fullName); // John Smith, lo stato di user è protetto
```

> Nell'esempio in linea `(*)` la proprietà `admin.fullName` ha un getter nel prototype `user`, quindi viene invocato. In linea `(**)` la proprietà ha un setter nel prototype, che viene quindi invocato.

Reviewer 2: ```suggestion
Nell'esempio, alla linea `(*)` la proprietà `admin.fullName` ha un getter nel prototype `user`, quindi viene invocato. E alla linea `(**)` la proprietà ha un setter nel prototype, che viene quindi invocato.
```

> Nell'esempio in linea `(*)` la proprietà `admin.fullName` possiede un getter nel prototype `user`, quindi viene invocato. In linea `(**)` la proprietà ha un setter nel prototype, che viene quindi invocato.

Reviewer: ```suggestion
Nell'esempio in linea `(*)` la proprietà `admin.fullName` ha un getter nel prototype `user`, quindi viene invocato. In linea `(**)` la proprietà ha un setter nel prototype, che viene quindi invocato.
```
Metterei un piu' sobrio *ha*. Che ne pensi?

## Il valore di "this"

> Dall'esempio sopra potrebbe sorgere una domanda interessante: qual'è il valore di `this` all'interno `set fullName(value)`? Dove vengono scritte le proprietà `this.name` e `this.surname`: in `user` o `admin`?

Reviewer: ```suggestion
Dall'esempio sopra potrebbe sorgere una domanda interessante: qual è il valore di `this` all'interno `set fullName(value)`? Dove vengono scritte le proprietà `this.name` e `this.surname`: in `user` o `admin`?
```

La risposta è semplice: `this` non viene influenzato dai prototype.

> **Non ha importanza dove viene trovato il metodo: nell'oggetto o in un suo prototupe. Quando invochiamo un metodo, `this` fa sempre riferimento all'oggetto che precede il punto.**

Reviewer: ```suggestion
**Non ha importanza dove viene trovato il metodo: nell'oggetto o in un suo prototype. Quando invochiamo un metodo, `this` fa sempre riferimento all'oggetto che precede il punto.**
```

Quindi, l'invocazione del setter `admin.fullName=` utilizza `admin` come `this`, non `user`.

> Questo è molto importante, poiché potremmo avere un oggetto molto grande con molti metodi, e avere diversi oggetti che ereditano da esso. Quando gli oggetti che ereditano, eseguono un metodo ereditato, andranno a modificare solamente il loro stato, non quello dell'oggetto principale.

Reviewer: ```suggestion
Questo è molto importante, poiché potremmo avere un oggetto molto grande con molti metodi, e avere diversi oggetti che ereditano da esso. Quando gli oggetti che ereditano eseguono un metodo ereditato, andranno a modificare solamente il loro stato, non quello dell'oggetto principale.
```

> Ad esempio, qui `animal` rappresenta un "memorizzato di metodi", che `rabbit` utilizza.

Reviewer: ```suggestion
Ad esempio, qui `animal` rappresenta un "archivio di metodi", che `rabbit` utilizza.
```

  __proto__: animal
};

// modifica rabbit.isSleeping
rabbit.sleep();

alert(rabbit.isSleeping); // true
alert(animal.isSleeping); // undefined (non esiste questa proprietà nel prototype)
```

Il risultato:

![](proto-animal-rabbit-walk-3.svg)

> Se avessimo altri oggetti, come `bird`, `snake`, etc., che ereditano da `animal`, avrebbero a loro volta accesso ai metodi di `animal`. In ogni caso, `this` all'interno della chiamata, farebbere riferimento all'oggetto corrispondente, che viene valutato al momento dell'invocazione (appena prima del punto), e non ad `animal`. Quindi quando scriviamo dati utilizzando `this`, questi verranno memorizzati nell'oggetto corrispondente.

Reviewer: ```suggestion
Se avessimo altri oggetti, come `bird`, `snake`, etc., che ereditano da `animal`, avrebbero a loro volta accesso ai metodi di `animal`. In ogni caso, `this` all'interno della chiamata farebbe riferimento all'oggetto corrispondente, che viene valutato al momento dell'invocazione (appena prima del punto), e non ad `animal`. Quindi quando scriviamo dati utilizzando `this`, questi verranno memorizzati nell'oggetto corrispondente.
```

> Come risultato otteniamo che dei metodi condivisi, mentre lo stato degli oggetti non lo è.

Reviewer: ```suggestion
Come risultato i metodi sono condivisi, mentre lo stato degli oggetti non lo è.
```

};

*!*
// Object.keys ritorna solamente le chiavi
alert(Object.keys(rabbit)); // jumps
*/!*

*!*

> // il ciclo for..in itera sia le proprietà di rabbi, che quelle ereditate da animal

Reviewer: ```suggestion
// il ciclo for..in itera sia le proprietà di rabbit, che quelle ereditate da animal
```

for(let prop in rabbit) alert(prop); // jumps, then eats
*/!*
```

> Se questo non è ciò che ci aspettiamo, e voglia escludere le proprietà ereditate, esiste un metodo integrato [obj.hasOwnProperty(key)](mdn:js/Object/hasOwnProperty): ritorna `true` se `obj` possiede la propria proprietà `key` (non ereditata).

Reviewer 2: ```suggestion
Se questo non è ciò che ci aspettiamo, e vogliamo escludere le proprietà ereditate, esiste un metodo integrato [obj.hasOwnProperty(key)](mdn:js/Object/hasOwnProperty): ritorna `true` se `obj` possiede la propria proprietà `key` (non ereditata).
```

> Se questo non è ciò che ci aspettiamo, e voglia escludere le proprietà ereditate, esiste un metodo integrato [obj.hasOwnProperty(key)](mdn:js/Object/hasOwnProperty): ritorna `true` se `obj` possiede la proprietà `key` come non ereditata (propria).

Reviewer: ```suggestion
Se questo non è ciò che ci aspettiamo, e voglia escludere le proprietà ereditate, esiste un metodo integrato [obj.hasOwnProperty(key)](mdn:js/Object/hasOwnProperty): ritorna `true` se `obj` possiede la propria proprietà `key`  (non ereditata).
```

> Quindi possiamo filtrare le proprietà ereditate (o farci qualcos altro):

Reviewer: ```suggestion
Quindi possiamo filtrare le proprietà ereditate (o farci qualcos'altro):
```

La risposta è semplice: la proprietà è *non enumerable*. Come tutte le altre proprietà di `Object.prototype`, possiedono la flag `enumerable:false`. Quindi `for..in` elenca solamente le proprietà enumerable. Questo è il motivo per cui le proprietà di `Object.prototype` non vengono elencate.

```smart header="Quasi tutti gli altri metodi getter key-value ignorano le proprietà ereditate"
Quasi tutti gli altri metodi getter key-value, come `Object.keys`, `Object.values` e cosi via, ignorano le proprietà ereditate.

Questi metodi lavorano solamente sull'oggetto stesso. Le proprietà di prototype *non* vengono prese in considerazione.
```

## Riepilogo

- In JavaScript, tutti gli oggetti possiedono una proprietà nascosta `[[Prototype]]` che può essere il riferimento ad un altro oggetto, oppure `null`.
- Possiamo utilizzare `obj.__proto__` per accedervi (una proprietà getter/setter storica, ci sono altri modi che vederemo presto).
- L'oggetto a cui fa riferimento `[[Prototype]]` viene chiamato "prototype".
- Se vogliamo leggere una proprietà di `obj` o invocare un metodo, ma questo non esiste, allora JavaScript andrà a cercarlo nel prototype.

> - Le operazioni di scrittrua/rimozione agiscono direttamente nell'oggetto, non utilizzano il prototype (assumendo che questa sia una proprietà e non un setter).

Reviewer: ```suggestion
- Le operazioni di scrittura/rimozione agiscono direttamente sull'oggetto, non utilizzano il prototype (assumendo che questa sia una proprietà e non un setter).
```
